asapdiscovery-data/asapdiscovery/data/schema_v2/ligand.py

```diff
@@ -333,6 +356,102 @@ def set_expansion(
             parent=parent, provenance=provenance
         )
 
+    @property
+    def canonical_tautomer(self) -> "Ligand":
+        """
+        Get the canonical tautomer of the ligand.
+        Not necessarily the most physiologically relevant tautomer, but helpful for comparing ligands.
+        """
+        mol = self.to_oemol()
+        canonical_tautomer = oechem.OEGraphMol()
+        if oequacpac.OEGetUniqueProtomer(canonical_tautomer, mol):
+            return Ligand.from_oemol(
+                compound_name=self.compound_name,
+                mol=canonical_tautomer,
+                expansion_tag=StateExpansionTag.from_parent(
+                    parent=self,
+                    provenance={
+                        "expander": "oequacpac.OEGetUniqueProtomer",
+                        "oechem": oechem.OEChemGetVersion(),
+                        "quacpac": oequacpac.OEQuacPacGetVersion(),
+                    },
+                ),
+            )
+        else:
+            raise ValueError("Unable to generate canonical tautomer")
+
+    @property
+    def has_defined_stereo(self) -> bool:
+        """
+        Check if the ligand has defined stereochemistry
+        """
+        mol = self.to_oemol()
+        return mol.HasPerceived(oechem.OEPerceived_AtomStereo) or mol.HasPerceived(
+            oechem.OEPerceived_BondStereo
+        )
+
+    def is_chemically_equal(self, other: "Ligand") -> bool:
+        """
+        Check if the ligand is chemically equal to another ligand using the inchikey.
+        Both ligands must both have defined stereochemistry or both not have defined stereochemistry.
+        """
+        return (
+            self.fixed_inchikey == other.fixed_inchikey
+            and self.has_defined_stereo == other.has_defined_stereo
+        )
+
+    def is_stereoisomer(self, other: "Ligand"):
+        """
+        Check if the ligand is a possible stereoisomer of another ligand.
+        Returns False if the ligands are the same.
+        """
+        # First check if molecules are the same
+        if self.is_chemically_equal(other):
+            return False
+
+        return self.non_iso_smiles == other.non_iso_smiles
+
+    def is_tautomer(self, other: "Ligand"):
+        """
+        Check if the ligand is a tautomer of another ligand, collapsing protonation states.
+        Returns False if the ligands are the same or stereoisomers.
+        """
+        # First check if molecules are the same or just a stereoisomer
+        if self.is_chemically_equal(other) or self.is_stereoisomer(other):
+            return False
+        return self.canonical_tautomer.is_chemically_equal(other.canonical_tautomer)
+
+    def is_related(self, other: "Ligand"):
```

Comment:
I think this is a bit too vague, stereoisomerism, tautomerism etc are all relationships in the strict definition of the word, seems to return `False`  for these cases. What types of relationships is this trying to catch?

Reply:
yeah this is confusing but I'm trying to catch the case where the ligand is a tautomer of a stereoisomer - so the ligands are clearly related to each other, but not through either one of the simpler relationships. Ligands can either be tautomers or stereoisomers, but if they are both, than neither of the `is_tautomer` or `is_stereoisomer` functions will be True. I could change it to be "is_distantly_related" or "is_tautomer_of_stereoisomer" or something.

Reply:
Okay that makes sense, I would go with the very explicit `is_tautomer_of_stereoisomer` or something.  Works both ways right as well?

Reply:
i decided to push this functionality into the "get relationship" function since the logic is a bit hard to wrap your mind around, maybe we can come back to it

Reply:
Yep fair enough. :)